### ISSUE TITLE
non-counting Select queries running far far far slower or never completing at all than equivalent counting Select

### ISSUE THREAD

**Issue by ebremer** [openlink/virtuoso-opensource#924]
Large triple store of 8,303,803,147 of PubMed publication data.  Server 64 cores and 1TB RAM on Azure with Ubuntu 18. Latest release of Virtuoso 7.

This query runs in about 30 seconds and yields a count of 8:
```
prefix : <https://pubmed.ncbi.nlm.nih.gov/ns/>
select count (*) 
where {?AffiliationInfo :Affiliation ?Affiliation 
filter(!isnumeric(?Affiliation)&&contains(?Affiliation, "SUNNY"))}
```

This query has been running for 25 minutes so far.
```
prefix : <https://pubmed.ncbi.nlm.nih.gov/ns/>
select * 
where {?AffiliationInfo :Affiliation ?Affiliation 
filter(!isnumeric(?Affiliation)&&contains(?Affiliation, "SUNNY"))}
```
Why is query #2 running so slow?  I have variations of this that are more complicated, but this is the simplest.

**Comment by ebremer** [comment]
```
Version 07.20.3229-pthreads for Linux as of Aug 15 2020
```

query #2 appears to have "died"  CPU went from 100% and dropped to 0%.  From browser point-of-view, it looks like it is still running.

```
KeepAliveTimeout = 3600
MaxQueryExecutionTime = 3600
```

**Comment by ebremer** [comment]
I ran the same query from isql and it returned results withing 300 seconds.  I'm not sure why the web interface is failing.

**Comment by ebremer** [comment]
I increased timeouts in firefox using --> `about:config`
increased default values on the following to match longest running query time:
```
network.http.tcp_keepalive.long_lived_idle_time
network.tcp.keepalive.idle_time
network.http.connection-timeout
```

Virtuoso now returning results in firefox for above query in same time-frame as isql.

**Comment by HughWilliams** [comment]
@ebremer: Can you please provide the [SPARQL query compilation report](https://community.openlinksw.com/t/generate-sparql-query-compilation-report-from-a-virtuoso-sparql-endpoint/1046)s for both queries for analysis as to why the second none-count query runs so slow, as detailed in the documentation link ?

After bulk loading your datasets, was the Virtuoso server restarted or the scheduler reactivated, to reenable incremental FT indexing, as indicated in the [RDF Bulk Loader](http://vos.openlinksw.com/owiki/wiki/VOS/VirtBulkRDFLoader) documentation at step 6 ?

**Comment by ebremer** [comment]
`scheduler_interval(10);` throws an error:
```
SQL> checkpoint;

Done. -- 4021 msec.
SQL> checkpoint_interval(120);

Done. -- 1 msec.
SQL> scheduler_interval(10);

*** Error 42001: [Virtuoso Driver][Virtuoso Server]SR185: Undefined procedure DB.DBA.scheduler_interval.
at line 7 of Top-Level:
scheduler_interval(10)
SQL>
```

**Comment by ebremer** [comment]
Query Number 1
=================================================

```
SPARQL query after parsing, optimization, and converting back into SPARQL

SELECT ( COUNT(  1))
 FROM <ftp://ftp.ncbi.nlm.nih.gov/pubmed/>
 WHERE {  ?AffiliationInfo <https://pubmed.ncbi.nlm.nih.gov/ns/Affiliation> ?Affiliation .
    FILTER ( !( isnumeric( ?Affiliation)) )
    FILTER ( contains( ?Affiliation, "SUNNY") ) }

SPARQL query translated to SQL
For security reasons, code responsible for graph-level security is not generated and some account-specific data are intentionally made wrong.

SELECT  COUNT (
     1) AS "callret-0"
FROM DB.DBA.RDF_QUAD AS "s_5_1_t0"
WHERE
  "s_5_1_t0"."G" = __i2idn ( __bft( 'ftp://ftp.ncbi.nlm.nih.gov/pubmed/' , 1))
  AND 
  "s_5_1_t0"."P" = __i2idn ( __bft( 'https://pubmed.ncbi.nlm.nih.gov/ns/Affiliation' , 1))
  AND 
  not ( isnumeric ( __ro2sq ( "s_5_1_t0"."O")))
  AND 
  rdf_contains_impl ( __ro2lo ( "s_5_1_t0"."O"),  DB.DBA.RDF_MAKE_LONG_OF_SQLVAL ( 'SUNNY' ))
OPTION (QUIETCAST)

SQL execution plan

  { 
    Precode:
      0: DB.DBA.RDF_MAKE_LONG_OF_SQLVAL$27 := Call DB.DBA.RDF_MAKE_LONG_OF_SQLVAL (<c SUNNY>)
      7: BReturn 0
    fork {
        RDF_QUAD   2.8e+07 rows(s_5_1_t0.O$30)
         inlined  P =  IRI_ID"...Affiliation"  G =  IRI_ID"...pubmed/" 
        END Node
        After test:
        0: __ro2sq$32 := Call __ro2sq (s_5_1_t0.O$30)
        5: isnumeric$34 := Call isnumeric (__ro2sq$32)
        10: if ( 0  = isnumeric$34) then 14 else 29 unkn 29
        14: __ro2lo$36 := Call __ro2lo (s_5_1_t0.O$30)
        19: rdf_contains_impl$38 := Call rdf_contains_impl (__ro2lo$36, DB.DBA.RDF_MAKE_LONG_OF_SQLVAL$27)
        24: if ( 0  = rdf_contains_impl$38) then 29 else 28 unkn 29
        28: BReturn 1
        29: BReturn 0
        
        After code:
        0:  sum callret-0$43 1 set no set_ctr$40
        5: BReturn 0
      }
    Select (callret-0$43)
  }

Internal optimizer data
These data are primarily for OpenLink support, to get additional details about the query processing.

	QUERY: (line 5) REQUEST TOP NODE (SELECT result-mode):
  RETVALS: ARRAY OF NODES with 1 children: {
    (line 4) FUNCALL:
      FUNCTION NAME: UNAME `SPECIAL::bif:COUNT'
      AGGREGATE MODE: LONG 1
      ARGUMENT: (line 4) LITERAL:
        VALUE: LONG 1
        DATATYPE: UNAME `http://www.w3.org/2001/XMLSchema#integer'
   }
  RETVALS SELECT ID: STRING `s_5_1'
  SOURCES: ARRAY OF NODES with 1 children: {
    (line 3) DATA SOURCE: FROM (default)
      IRI: UNAME `ftp://ftp.ncbi.nlm.nih.gov/pubmed/'
      EXPN: (line 3) QNAME:
        IRI: UNAME `ftp://ftp.ncbi.nlm.nih.gov/pubmed/'
   }
  PATTERN: (line 5) GRAPH PATTERN:WHERE gp
    MEMBERS: ARRAY OF NODES with 1 children: {
      (line 4) TRIPLE:
        GRAPH: (line 3) QNAME:
          IRI: UNAME `ftp://ftp.ncbi.nlm.nih.gov/pubmed/'
        SUBJECT: (line 4) VARIABLE: notNULL+ reference+ (subject)
          NAME: UNAME `AffiliationInfo'	SELECT ID: STRING `s_5_1'	TABLE ID: STRING `s_5_1_t0'	EQUIV: LONG 0
        PREDICATE: (line 4) QNAME:
          IRI: UNAME `https://pubmed.ncbi.nlm.nih.gov/ns/Affiliation'
        OBJECT: (line 4) VARIABLE: notNULL+ (object)
          NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: STRING `s_5_1_t0'	EQUIV: LONG 1
        SELECT ID: STRING `s_5_1'
        TABLE ID: STRING `s_5_1_t0'
     }
    FILTERS: ARRAY OF NODES with 2 children: {
      (line 4) OPERATOR EXPRESSION (boolean operation 'NOT'):
        LEFT: (line 4) BUILT-IN CALL:isnumeric
          ARGUMENT: ARRAY OF NODES with 1 children: {
            (line 4) VARIABLE: notNULL
              NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: LONG 0	EQUIV: LONG 1
           }
      (line 4) BUILT-IN CALL:contains
        ARGUMENT: ARRAY OF NODES with 2 children: {
          (line 4) VARIABLE: notNULL
            NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: LONG 0	EQUIV: LONG 1
          (line 4) LITERAL:
            VALUE: STRING `SUNNY'
         }
     }
    SELECT ID: STRING `s_5_1'
    EQUIVS: 0 1

EQUIVS:
#0: ( 0 subv (0 bindings, 0 nest.opt.), 0 recv, 1 gspo, 0 const, 0 opt, 0 subq: AffiliationInfo in s_5_1_t0; notNULL reference)
#1: ( 0 subv (0 bindings, 0 nest.opt.), 0 recv, 1 gspo, 2 const, 0 opt, 0 subq: Affiliation in s_5_1_t0 s_5_1 s_5_1; notNULL)
```

QUERY Number 2
=========================
```
SPARQL query after parsing, optimization, and converting back into SPARQL

SELECT  ?AffiliationInfo
 ?Affiliation
 FROM <ftp://ftp.ncbi.nlm.nih.gov/pubmed/>
 WHERE {  ?AffiliationInfo <https://pubmed.ncbi.nlm.nih.gov/ns/Affiliation> ?Affiliation .
    FILTER ( !( isnumeric( ?Affiliation)) )
    FILTER ( contains( ?Affiliation, "SUNNY") ) }

SPARQL query translated to SQL
For security reasons, code responsible for graph-level security is not generated and some account-specific data are intentionally made wrong.

SELECT __id2in ( "s_5_1_t0"."S") AS "AffiliationInfo",
  __ro2sq ( "s_5_1_t0"."O") AS "Affiliation"
FROM DB.DBA.RDF_QUAD AS "s_5_1_t0"
WHERE
  "s_5_1_t0"."G" = __i2idn ( __bft( 'ftp://ftp.ncbi.nlm.nih.gov/pubmed/' , 1))
  AND 
  "s_5_1_t0"."P" = __i2idn ( __bft( 'https://pubmed.ncbi.nlm.nih.gov/ns/Affiliation' , 1))
  AND 
  not ( isnumeric ( __ro2sq ( "s_5_1_t0"."O")))
  AND 
  rdf_contains_impl ( __ro2lo ( "s_5_1_t0"."O"),  DB.DBA.RDF_MAKE_LONG_OF_SQLVAL ( 'SUNNY' ))
OPTION (QUIETCAST)

SQL execution plan

  { 
    Precode:
      0: DB.DBA.RDF_MAKE_LONG_OF_SQLVAL$27 := Call DB.DBA.RDF_MAKE_LONG_OF_SQLVAL (<c SUNNY>)
      7: BReturn 0
    RDF_QUAD   2.8e+07 rows(s_5_1_t0.O$31, s_5_1_t0.S$30)
     inlined  P =  IRI_ID"...Affiliation"  G =  IRI_ID"...pubmed/" 
    END Node
    After test:
    0: __ro2sq$33 := Call __ro2sq (s_5_1_t0.O$31)
    5: isnumeric$35 := Call isnumeric (__ro2sq$33)
    10: if ( 0  = isnumeric$35) then 14 else 29 unkn 29
    14: __ro2lo$37 := Call __ro2lo (s_5_1_t0.O$31)
    19: rdf_contains_impl$39 := Call rdf_contains_impl (__ro2lo$37, DB.DBA.RDF_MAKE_LONG_OF_SQLVAL$27)
    24: if ( 0  = rdf_contains_impl$39) then 29 else 28 unkn 29
    28: BReturn 1
    29: BReturn 0
    
    After code:
    0: AffiliationInfo$41 := Call __id2in (s_5_1_t0.S$30)
    5: Affiliation$43 := Call __ro2sq (s_5_1_t0.O$31)
    10: BReturn 0
    Select (AffiliationInfo$41, Affiliation$43)
  }

Internal optimizer data
These data are primarily for OpenLink support, to get additional details about the query processing.

	QUERY: (line 5) REQUEST TOP NODE (SELECT result-mode):
  RETVALS: ARRAY OF NODES with 2 children: {
    (line 5) VARIABLE: notNULL reference exported
      NAME: UNAME `AffiliationInfo'	SELECT ID: STRING `s_5_1'	TABLE ID: LONG 0	EQUIV: LONG 0
    (line 5) VARIABLE: notNULL exported
      NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: LONG 0	EQUIV: LONG 1
   }
  RETVALS SELECT ID: STRING `s_5_1'
  SOURCES: ARRAY OF NODES with 1 children: {
    (line 3) DATA SOURCE: FROM (default)
      IRI: UNAME `ftp://ftp.ncbi.nlm.nih.gov/pubmed/'
      EXPN: (line 3) QNAME:
        IRI: UNAME `ftp://ftp.ncbi.nlm.nih.gov/pubmed/'
   }
  PATTERN: (line 5) GRAPH PATTERN:WHERE gp
    MEMBERS: ARRAY OF NODES with 1 children: {
      (line 4) TRIPLE:
        GRAPH: (line 3) QNAME:
          IRI: UNAME `ftp://ftp.ncbi.nlm.nih.gov/pubmed/'
        SUBJECT: (line 4) VARIABLE: notNULL+ reference+ exported (subject)
          NAME: UNAME `AffiliationInfo'	SELECT ID: STRING `s_5_1'	TABLE ID: STRING `s_5_1_t0'	EQUIV: LONG 0
        PREDICATE: (line 4) QNAME:
          IRI: UNAME `https://pubmed.ncbi.nlm.nih.gov/ns/Affiliation'
        OBJECT: (line 4) VARIABLE: notNULL+ exported (object)
          NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: STRING `s_5_1_t0'	EQUIV: LONG 1
        SELECT ID: STRING `s_5_1'
        TABLE ID: STRING `s_5_1_t0'
     }
    FILTERS: ARRAY OF NODES with 2 children: {
      (line 4) OPERATOR EXPRESSION (boolean operation 'NOT'):
        LEFT: (line 4) BUILT-IN CALL:isnumeric
          ARGUMENT: ARRAY OF NODES with 1 children: {
            (line 4) VARIABLE: notNULL exported
              NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: LONG 0	EQUIV: LONG 1
           }
      (line 4) BUILT-IN CALL:contains
        ARGUMENT: ARRAY OF NODES with 2 children: {
          (line 4) VARIABLE: notNULL exported
            NAME: UNAME `Affiliation'	SELECT ID: STRING `s_5_1'	TABLE ID: LONG 0	EQUIV: LONG 1
          (line 4) LITERAL:
            VALUE: STRING `SUNNY'
         }
     }
    SELECT ID: STRING `s_5_1'
    EQUIVS: 0 1

EQUIVS:
#0: ( 0 subv (0 bindings, 0 nest.opt.), 0 recv, 1 gspo, 1 const, 0 opt, 0 subq: AffiliationInfo in s_5_1_t0 s_5_1; notNULL reference exported)
#1: ( 0 subv (0 bindings, 0 nest.opt.), 0 recv, 1 gspo, 3 const, 0 opt, 0 subq: Affiliation in s_5_1_t0 s_5_1 s_5_1 s_5_1; notNULL exported)
```

**Comment by HughWilliams** [comment]
@ebremer: The `scheduler_interval()` function has existing in Virtuoso for as long as I have worked with it, which is over 20 yrs now, and works with my local Virtuoso 7.2 open source build:

```
SQL> scheduler_interval(10);

Done. -- 2 msec.
SQL>
```
my version and gitid being:
```
virtuoso-t -?
Virtuoso Open Source Edition (Column Store) (multi threaded)
Version 7.2.6-rc1.3230-pthreads as of Oct 14 2019 (0e746c7ab)
```

What is your version and gitid ?

**Comment by ebremer** [comment]
```
OpenLink Virtuoso Interactive SQL (Virtuoso)
Version 07.20.3229 as of Aug 15 2020
Type HELP; for help and EXIT; to exit.
Connected to OpenLink Virtuoso
Driver: 07.20.3229 OpenLink Virtuoso ODBC Driver
SQL> scheduler_interval(10);

*** Error 42001: [Virtuoso Driver][Virtuoso Server]SR185: Undefined procedure DB.DBA.scheduler_interval.
at line 1 of Top-Level:
scheduler_interval(10)
```

**Comment by HughWilliams** [comment]
@ebremer: I stand corrected, as although `checkpoint_interval()` has always existed, the `scheduler_interval()` function was added in 2019, thus I can only assume you are running a binary built from a source archive from before 2019 (possible the current `stable/7`) branch which is why I wanted the gitid from the `virtuoso-t -?` output which would confirm this, thus can you please provide this output for your binary?

You would need to build from the `develop/7` branch to get the `scheduler_interval` function, which we would recommend as it has many fixes / improvements not in the current `stable/7`  branch, and is being prepared currently to become the new `stable/7` branch.

**Comment by ebremer** [comment]
```
Virtuoso Open Source Edition (Column Store) (multi threaded)
Version 7.2.5.3229-pthreads as of Aug 15 2020 (000000)
Compiled for Linux (x86_64-pc-linux-gnu)
Copyright (C) 1998-2018 OpenLink Software
```

I will build the `develop/7` branch if you are asserting it is stable enough for production.

**Comment by HughWilliams** [comment]
@ebremer: Your gitid is `000000` thus did you download a zip or tar archive of the `stable/7` branch from the git repo rather then cloning it, as these would not include the files needed to generate the gitid, thus we always recommend cloning the git repo which will then result in the necessary files being generated.

So we would recommend cloning the `develop/7` branch which is the default, and build and repeat the running of the two queries against `develop/7` build, and if the problem persists provide the query compilation reports from this binary please.

Note: when updating the binary, ensure a `checkpoint` is run before stopping the database to commit pending transactions to the database file such that the `trx` file is empty before starting with the new binary.

**Comment by ebremer** [comment]
Yes, I used the zip file for 7.2.5.1 release.  I just rebuilt both of my triple stores and built them on `develop/7` and applied fix from #742 which applied also to my other issue of #923 .  I had tested `develop/7` without this fix and my server crashed on startup but works fine now with `develop/7` with fix from #742 .  I will re-test and let you know what happens.  Thank you for your help!
